<a href="https://colab.research.google.com/github/aplneto/IF1014/blob/main/07_Artificial_Neural_Networks.ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>

# <center> Missão 7 </center>

#Equipe:
Antônio Paulino - apln2@cin.ufpe.br
Ailton Rodrigues - ajr@cin.ufpe.br
Douglas Tavares - dtrps@cin.ufpe.br

#Parte 1: As redes neurais artificiais estão mais do que nunca em alta devido aos seus benefícios em inúmeras áreas e pelo crescente volume de dados. Em um problema a ser tratado com métodos de aprendizagem de máquina, seria melhor optar por modelos já vistos na disciplina ou partir diretamente para modelos de redes neurais? por que?


 
Como um bom Cientista de Dados/Engenheiro de Aprendizado de Máquina, você deve estar bem ciente do arsenal que possui e qual arma deve se usada para garantir o resultado máximo.
 
Partindo dessa premissa, existem 3 razões principais para partir diretamente para modelos de redes neurais:
 
i.               Há disponibilização de muitos dados
 
a.    As redes neurais exigem muito mais dados do que os algoritmos de aprendizado de máquina tradicionais para fazer seu trabalho bem; logo, se existem muitos dados disponíveis a opção é redes neurais.
 
ii.             Poder computacional
 
a.    As redes neurais exigem um processo de treinamento e para isso é necessário ter um poder computacional alto pra poder executar os experimentos. Como atualmente o poder computacional dobra a cada dois anos, então isso não será empecilho para não usar redes neurais.
 
iii.            Algoritmo de retropropagação
 
a.    A retropropagação ajuda na formação de redes neurais artificiais. Quando as redes neurais artificiais se formam, os valores dos pesos são atribuídos aleatoriamente. O utilizador estabelece pesos aleatórios porque não está ciente dos valores corretos. Quando o valor é diferente da rede de alimentação esperada, considere-o como um erro. O algoritmo é definido para que o modelo altere os parâmetros cada vez que a saída não é a esperada. O erro tem uma relação com redes neurais artificiais. Assim, quando o parâmetro muda, o erro também muda até a rede neural encontrar a saída desejada através do cálculo do gradiente de descida
 
 
Ademais, possuem vantagens que ratificam o por quê das redes neurais estarem no auge e por serem uma “arma” predileta do arsenal:
 
i.               a capacidade de aprender e modelar relações não lineares e complexas, o que é realmente importante porque, na vida real, muitas das relações entre entradas e saídas são não lineares e também complexas.  Ademais, o crescente volume de dados propicia o treinamento de modelos mais robustos com o de aprendizado profundo.
 
ii.              As redes neurais podem generalizar - depois de aprender com as entradas iniciais e seus relacionamentos, pode inferir relacionamentos não vistos em dados não vistos também, fazendo com que o modelo generalize e preveja em dados não vistos.
 
iii.             Ao contrário de muitas outras técnicas de predição, a RNA não impõe nenhuma restrição às variáveis de entrada (por exemplo, como devem ser distribuídas). Além disso, muitos estudos mostraram que as RNAs podem modelar melhor dados com alta volatilidade e variância não constante, dada sua capacidade de aprender relações ocultas nos dados sem impor quaisquer relações fixas nos dados. Isso é algo muito útil na previsão de séries temporais financeiras (por exemplo, preços de ações) em que a volatilidade dos dados é muito alta.

## Instalação dos pacotes necessários

In [ ]:
!python3 -m pip install optuna

     |████████████████████████████████| 308 kB 20.9 MB/s 
     |████████████████████████████████| 209 kB 31.2 MB/s 
     |████████████████████████████████| 80 kB 6.9 MB/s 
     |████████████████████████████████| 75 kB 3.3 MB/s 
     |████████████████████████████████| 112 kB 39.3 MB/s 
     |████████████████████████████████| 144 kB 49.0 MB/s 
     |████████████████████████████████| 49 kB 6.8 MB/s 
  Created wheel for pyperclip: filename=pyperclip-1.8.2-py3-none-any.whl size=11136 sha256=efca9fb1c2cab5263bf13a5b7a6a46c08e8181651908cc5ed18569158b61ed1c
  Stored in directory: /root/.cache/pip/wheels/9f/18/84/8f69f8b08169c7bae2dde6bd7daf0c19fca8c8e500ee620a28
Successfully built pyperclip


# Análise Exploratória e preparação dos dados

In [ ]:
DATA_FOLDER = (
    'https://archive.ics.uci.edu/ml/machine-learning-databases/'
    'credit-screening/'
)

DATA_DESCRIPTION = DATA_FOLDER + 'crx.names'
DATA_SET = DATA_FOLDER + 'crx.data'

In [ ]:
import pandas
import numpy

aliases = [
  'Gender', 'Age', 'Debt', 'Married', 'BankCustomer', 'EducationLevel',
  'Ethnicity', 'YearsEmployed', 'PriorDefault', 'Employed', 'CreditScore',
  'DriversLicense', 'Citizen', 'ZipCode', 'Income', 'Approved'
]
data = pandas.read_csv(DATA_SET, names=aliases, na_values='?', header=None)
data.head()

Gender    Age   Debt Married  ... Citizen ZipCode Income  Approved
0      b  30.83  0.000       u  ...       g   202.0      0         +
1      a  58.67  4.460       u  ...       g    43.0    560         +
2      a  24.50  0.500       u  ...       g   280.0    824         +
3      b  27.83  1.540       u  ...       g   100.0      3         +
4      b  20.17  5.625       u  ...       s   120.0      0         +

[5 rows x 16 columns]

#### Remoção das variáveis Ethnicity (A7) e ZipCode (A14) por não exercerem influência na variável alvo

In [ ]:
# removing useless variables A7 (Ethnicity) and A14 (ZipCode)

data.drop(['Ethnicity', 'ZipCode'], axis=1, inplace=True)
data.head()

Gender    Age   Debt Married  ... DriversLicense Citizen  Income Approved
0      b  30.83  0.000       u  ...              f       g       0        +
1      a  58.67  4.460       u  ...              f       g     560        +
2      a  24.50  0.500       u  ...              f       g     824        +
3      b  27.83  1.540       u  ...              t       g       3        +
4      b  20.17  5.625       u  ...              f       s       0        +

[5 rows x 14 columns]

#### Separação das variáveis em contínuas e categóricas

In [ ]:
continuous = data.describe().columns
categorical = data.drop(list(continuous) + ['Approved'], axis=1).columns

print(continuous)
print(categorical)

Index(['Age', 'Debt', 'YearsEmployed', 'CreditScore', 'Income'], dtype='object')
Index(['Gender', 'Married', 'BankCustomer', 'EducationLevel', 'PriorDefault',
       'Employed', 'DriversLicense', 'Citizen'],
      dtype='object')


# Limpeza dos dados

## Modelo de regressão linear para completar dados continuos ausentes

Variáveis continuas ausentes serão preenchidas por valores previstos a partir de um modelo de regressão linear construído a partir da coluna com valores ausentes e da coluna com todos os valores mais fortemente correlacionada a ela

In [ ]:
continuous_columns_missing_values = []

for column in continuous:
  if data[column].isnull().sum() > 0:
    continuous_columns_missing_values.append(column)

print(continuous_columns_missing_values)

['Age']


In [ ]:
most_correlated_columns = {}
candidates = [
  x for x in continuous if x not in continuous_columns_missing_values
]
for column in continuous_columns_missing_values:
  most_correlated_columns[column] = max(
      candidates, key=lambda x: abs(data[x].corr(data[column]))
  )

In [ ]:
from sklearn.linear_model import LinearRegression

In [ ]:
prediction_models = {}

for pair in most_correlated_columns.items():
  rows = data[~data[list(pair)].isnull().any(axis=1)][list(pair)]
  y = rows[pair[0]]
  x = rows[pair[1]]
  lr = LinearRegression()
  lr.fit(x.values.reshape(-1, 1), y)
  d = pandas.DataFrame(data= {
      'value' : lr.predict(data[pair[1]].values.reshape(-1, 1))
  })
  data[pair[0]] = numpy.where(data[column].isna(), d['value'], data[column])

data[continuous].isna().sum()

Age              0
Debt             0
YearsEmployed    0
CreditScore      0
Income           0
dtype: int64

# Codificação ortogonal

Valores categóricos foram codificados no formato one hot encoding.

Os valores ausentes foram completados usando um algoritmo de árvore de decisão.

In [ ]:
categorical_columns_missing_values = [
  p[0] for p in dict(data[categorical].isna().sum() > 0).items() if p[1]
]
complete_data = data.dropna()
print(categorical_columns_missing_values)

['Gender', 'Married', 'BankCustomer', 'EducationLevel']


In [ ]:
from sklearn.preprocessing import LabelEncoder
from collections import defaultdict

label_dict = defaultdict(LabelEncoder)
complete_data = complete_data.apply(
    lambda x: label_dict[x.name].fit_transform(x)
    if x.name in list(categorical) + ['Approved']
    else x
)

In [ ]:
from sklearn.tree import DecisionTreeClassifier

In [ ]:
trees = {}
X = complete_data.drop(categorical_columns_missing_values, axis=1)
for column in categorical_columns_missing_values:
  Y = complete_data[column]
  tree = DecisionTreeClassifier(
      max_leaf_nodes=Y.nunique(), random_state=2**Y.nunique()
  )
  trees[column] = tree
  tree.fit(X.values, Y.values)

In [ ]:
for column in trees:
  tree = trees[column]
  encoder = label_dict[column]
  d = pandas.DataFrame(data= {
      'value' : encoder.inverse_transform(
          tree.predict(
          data.drop(categorical_columns_missing_values, axis=1).apply(
                  lambda x: label_dict[x.name].fit_transform(x)
                  if x.name in list(categorical) + ['Approved']
                  else x
          ).values
        )
      )
    }
  )

  data[column] = numpy.where(data[column].isna(), d['value'], data[column])

In [ ]:
from sklearn.preprocessing import OneHotEncoder

In [ ]:
onehotencoders = defaultdict(OneHotEncoder)

new_categories = []

for column in categorical:
  encoder = onehotencoders[column]
  encoder.fit(data[column].values.reshape(-1, 1))
  arr = encoder.transform(data[column].values.reshape(-1, 1)).toarray()
  data.drop(column, axis=1, inplace=True)
  d = dict(zip(*[column+'_'+ c for c in encoder.categories_], zip(*arr)))
  for k in d:
    data[k] = d[k]
    new_categories.append(k)

data.head()
categorical = new_categories

In [ ]:
# encoder = OneHotEncoder()
# encoder.fit(data['Gender'].values.reshape(-1, 1))
# arr = encoder.transform(data['Gender'].values.reshape(-1, 1)).toarray()
# d = dict(zip(*encoder.categories_, zip(*arr)))
# t = dict(zip(*['data_'+ c for c in encoder.categories_], zip(*arr)))

In [ ]:
labels = data['Approved']
data.drop('Approved', axis=1, inplace=True)
X = data.apply(
    lambda x: label_dict[x.name].fit_transform(x)
    if x.name in categorical
    else x
)
print(X)

       Age    Debt  YearsEmployed  ...  Citizen_g  Citizen_p  Citizen_s
0    30.83   0.000           1.25  ...          1          0          0
1    58.67   4.460           3.04  ...          1          0          0
2    24.50   0.500           1.50  ...          1          0          0
3    27.83   1.540           3.75  ...          1          0          0
4    20.17   5.625           1.71  ...          0          0          1
..     ...     ...            ...  ...        ...        ...        ...
685  21.08  10.085           1.25  ...          1          0          0
686  22.67   0.750           2.00  ...          1          0          0
687  25.25  13.500           2.00  ...          1          0          0
688  17.92   0.205           0.04  ...          1          0          0
689  35.00   3.375           8.29  ...          1          0          0

[690 rows x 36 columns]


In [ ]:
Y = pandas.DataFrame(
    LabelEncoder().fit_transform(labels), columns=[labels.name]
)
print(Y)

     Approved
0           0
1           0
2           0
3           0
4           0
..        ...
685         1
686         1
687         1
688         1
689         1

[690 rows x 1 columns]


# Divisão das instâncias em treinamento e teste

In [ ]:
from sklearn.model_selection import train_test_split

In [ ]:
X_train, X_test, Y_train, Y_test = train_test_split(
    X.values, Y.values, test_size = 0.2, random_state = 4
)

In [ ]:
X_train = pandas.DataFrame(X_train, columns=X.columns)
X_test = pandas.DataFrame(X_test, columns=X.columns)
Y_train = pandas.DataFrame(Y_train, columns=Y.columns)
Y_test = pandas.DataFrame(Y_test, columns=Y.columns)

In [ ]:
print(X_train.shape)
print(Y_train.shape)
print(X_test.shape)
print(Y_test.shape)

(552, 36)
(552, 1)
(138, 36)
(138, 1)


# Transformação dos Dados

* Uma vez que as variáveis continuas possuem valores entre 0 e um determinado limite, estes serão normalizados entre os valores 0.0 e 1.0 para análise de diminuição de dimensionalidade

In [ ]:
X_train[continuous].describe()

Age        Debt  YearsEmployed  CreditScore         Income
count  552.000000  552.000000     552.000000   552.000000     552.000000
mean    31.375927    4.723342       2.222554     2.574275    1016.204710
std     11.791370    4.958010       3.342170     5.163208    5328.577631
min     13.750000    0.000000       0.000000     0.000000       0.000000
25%     22.500000    0.875000       0.165000     0.000000       0.000000
50%     28.448036    2.750000       1.000000     0.000000       5.500000
75%     37.520000    7.312500       2.551250     3.000000     462.250000
max     80.250000   28.000000      28.500000    67.000000  100000.000000

In [ ]:
X_test[continuous].describe()

Age        Debt  YearsEmployed  CreditScore        Income
count  138.000000  138.000000     138.000000   138.000000    138.000000
mean    32.437283    4.900254       2.226812     1.702899   1022.108696
std     12.172970    5.073769       3.376051     3.331796   4724.580194
min     15.170000    0.000000       0.000000     0.000000      0.000000
25%     23.080000    1.250000       0.125000     0.000000      0.000000
50%     29.585000    2.730000       0.750000     0.000000      1.000000
75%     39.107500    7.000000       2.750000     2.000000    200.000000
max     76.750000   25.210000      17.500000    20.000000  50000.000000

In [ ]:
from sklearn.preprocessing import MinMaxScaler

In [ ]:
train_scalers = defaultdict(MinMaxScaler)
test_scalers = defaultdict(MinMaxScaler)

for column in continuous:
  train_scaler = train_scalers[column]
  test_scaler = test_scalers[column]
  X_train[column] = train_scaler.fit_transform(X_train[column].values.reshape(-1, 1))
  X_test[column] = test_scaler.fit_transform(X_test[column].values.reshape(-1, 1))

# Redução da dimensionalidade

<!--

* https://www.analyticsvidhya.com/blog/2018/08/dimensionality-reduction-techniques-python/

 -->

## Principal component analysis

<!--

* https://www.datasklr.com/principal-component-analysis-and-factor-analysis/principal-component-analysis
* https://www.youtube.com/watch?v=FgakZw6K1QQ
* https://jmausolf.github.io/code/pca_in_python/

-->

In [ ]:
from sklearn.decomposition import PCA

In [ ]:
pca_train = PCA()
pca_train.fit(X_train[continuous])

PCA(copy=True, iterated_power='auto', n_components=None, random_state=None,
    svd_solver='auto', tol=0.0, whiten=False)

In [ ]:
numpy.cumsum(pca_train.explained_variance_ratio_)

array([0.49681538, 0.7953898 , 0.91164728, 0.96771056, 1.        ])

In [ ]:
pca_X_train = pandas.DataFrame(
    data = pca_train.transform(X_train[continuous]),
    columns = ['PC%d' % (i) for i in numpy.arange(pca_train.n_components_)+1]
)

max_column = numpy.argmax(numpy.cumsum(pca_train.explained_variance_ratio_) > 0.9) + 1
principal_components = pca_X_train.columns[:max_column:]

pca_X_train = pandas.concat(
    [pca_X_train[principal_components], X_train[categorical]],
    axis = 1
)

pca_X_train

PC1       PC2       PC3  ...  Citizen_g  Citizen_p  Citizen_s
0   -0.080758 -0.184270 -0.065782  ...        0.0        1.0        0.0
1   -0.216770 -0.018099  0.009423  ...        0.0        0.0        1.0
2   -0.189481 -0.012675  0.048359  ...        1.0        0.0        0.0
3    0.087857 -0.047823 -0.124984  ...        1.0        0.0        0.0
4   -0.130168  0.191575 -0.026154  ...        1.0        0.0        0.0
..        ...       ...       ...  ...        ...        ...        ...
547 -0.143830  0.060022  0.087901  ...        1.0        0.0        0.0
548  0.041393  0.023916  0.093961  ...        0.0        0.0        1.0
549 -0.167171 -0.040376 -0.002445  ...        1.0        0.0        0.0
550  0.028834  0.238163 -0.045819  ...        1.0        0.0        0.0
551  0.125133  0.276131 -0.029625  ...        1.0        0.0        0.0

[552 rows x 34 columns]

In [ ]:
pca_X_test = pandas.DataFrame(
    data = pca_train.transform(X_test[continuous]),
    columns = ['PC%d' % (i) for i in numpy.arange(pca_train.n_components_)+1]
)

principal_components = pca_X_test.columns[:max_column:]

pca_X_test = pandas.concat(
    [pca_X_test[principal_components], X_test[categorical]],
    axis = 1
)

pca_X_test

PC1       PC2       PC3  ...  Citizen_g  Citizen_p  Citizen_s
0    0.013034  0.154900  0.158099  ...        1.0        0.0        0.0
1    0.616798 -0.115678 -0.053012  ...        1.0        0.0        0.0
2   -0.141278  0.023545 -0.017648  ...        1.0        0.0        0.0
3    0.207157 -0.161917  0.414889  ...        0.0        0.0        1.0
4   -0.293477  0.043979  0.033666  ...        1.0        0.0        0.0
..        ...       ...       ...  ...        ...        ...        ...
133 -0.131200  0.060932  0.003077  ...        1.0        0.0        0.0
134 -0.245238  0.012432  0.017578  ...        1.0        0.0        0.0
135  0.305105 -0.170936 -0.040540  ...        1.0        0.0        0.0
136  0.237959  0.276361 -0.153109  ...        1.0        0.0        0.0
137 -0.108243  0.153372 -0.005270  ...        1.0        0.0        0.0

[138 rows x 34 columns]

In [ ]:
import optuna
from sklearn.neural_network import MLPClassifier

In [ ]:
(1,None)

(1, None)

In [ ]:
def model_factory(hidden_layers, activation_fn, _solver, l2_alpha, **args):
  return MLPClassifier(
      hidden_layer_sizes=hidden_layers,
      activation = activation_fn,
      solver = _solver,
      alpha = l2_alpha,
      **args
  )

def mlp_model_optimization_study(trial: optuna.trial.FixedTrial):
  insize = pca_X_train.values.shape[1] + 1
  outsize = 2
  number_of_layers = trial.suggest_int('number_of_layers', 1, 2)
  hl1 = trial.suggest_int('hidden_layer_1', insize*2, 100)
  hidden_layers = (
      (hl1, trial.suggest_int('hidden_layer_2', outsize, insize))
      if number_of_layers > 1
      else (hl1,)
  )
  activation_fn = trial.suggest_categorical(
      'activation_fn', ['identity', 'logistic', 'tanh', 'relu']
  )
  _solver = trial.suggest_categorical(
      '_solver', ['lbfgs', 'sgd', 'adam']
  )
  l2_alpha = trial.suggest_loguniform('l2_alpha', 0.1, 10)
  early_stopping = (
      trial.suggest_categorical('early_stopping', [False, True])
      if _solver != 'lbfgs'
      else False
  )
  epochs = (
      trial.suggest_int('epochs', 100, 500, 50)
      if _solver != 'lbfgs'
      else trial.suggest_int('epochs', 5000, 10000, 1000)
  )
  model = (
      model_factory(hidden_layers, activation_fn, _solver, l2_alpha,
                    early_stopping = True)
      if early_stopping
      else model_factory(hidden_layers, activation_fn, _solver, l2_alpha,
                         early_stopping = True,
                         max_iter=epochs)
  )
  model.fit(pca_X_train.values, Y_train.values.ravel())
  trial.set_user_attr(
      'training_score', model.score(pca_X_train.values, Y_train.values.ravel())
  )
  return model.score(pca_X_test.values, Y_test.values.ravel())
  
#adam e sgd

In [ ]:
study = optuna.study.create_study(
    study_name='MLP_study', direction='maximize'
)
study.optimize(mlp_model_optimization_study, n_trials=100)

[I 2021-11-09 23:40:09,040] A new study created in memory with name: MLP_study
[I 2021-11-09 23:40:09,138] Trial 0 finished with value: 0.8333333333333334 and parameters: {'number_of_layers': 1, 'hidden_layer_1': 88, 'activation_fn': 'relu', '_solver': 'adam', 'l2_alpha': 2.361587584583161, 'early_stopping': False, 'epochs': 250}. Best is trial 0 with value: 0.8333333333333334.
[I 2021-11-09 23:40:09,451] Trial 1 finished with value: 0.8260869565217391 and parameters: {'number_of_layers': 2, 'hidden_layer_1': 94, 'hidden_layer_2': 28, 'activation_fn': 'identity', '_solver': 'lbfgs', 'l2_alpha': 0.38623342815442774, 'epochs': 5000}. Best is trial 0 with value: 0.8333333333333334.
[I 2021-11-09 23:40:11,644] Trial 2 finished with value: 0.782608695652174 and parameters: {'number_of_layers': 2, 'hidden_layer_1': 72, 'hidden_layer_2': 4, 'activation_fn': 'tanh', '_solver': 'lbfgs', 'l2_alpha': 2.672146512157626, 'epochs': 5000}. Best is trial 0 with value: 0.8333333333333334.
[I 2021-11-09

In [ ]:
for i, trial in enumerate(study.trials):
  print('Trial %i' % i)
  print('Acurácia de treinamento:',trial.user_attrs['training_score'])

Trial 0
Acurácia de treinamento: 0.8713768115942029
Trial 1
Acurácia de treinamento: 0.8858695652173914
Trial 2
Acurácia de treinamento: 0.9402173913043478
Trial 3
Acurácia de treinamento: 0.5489130434782609
Trial 4
Acurácia de treinamento: 0.5489130434782609
Trial 5
Acurácia de treinamento: 0.802536231884058
Trial 6
Acurácia de treinamento: 0.9329710144927537
Trial 7
Acurácia de treinamento: 0.8623188405797102
Trial 8
Acurácia de treinamento: 0.9166666666666666
Trial 9
Acurácia de treinamento: 0.8641304347826086
Trial 10
Acurácia de treinamento: 0.875
Trial 11
Acurácia de treinamento: 0.8460144927536232
Trial 12
Acurácia de treinamento: 0.8768115942028986
Trial 13
Acurácia de treinamento: 0.8786231884057971
Trial 14
Acurácia de treinamento: 0.8786231884057971
Trial 15
Acurácia de treinamento: 0.8586956521739131
Trial 16
Acurácia de treinamento: 0.8206521739130435
Trial 17
Acurácia de treinamento: 0.7916666666666666
Trial 18
Acurácia de treinamento: 0.8731884057971014
Trial 19
Acurácia

# <center>Considerações Finais </center>

  O algoritmo de redes neurais foi treinado com 4 funções de ativação: *identity*, *logistic*, *tanh* e *relu*. Como também 3 *solvers*: lbfgs, sgd, adam. Foi utilizado também a regulariação nas redes neurais para prevenir o *overfitting*, porém no *Scikit learn* só tinha disponível a regularização L2, em pesquisas realizadas a regularização L2 trabalha melhor no intervalo de 0.1 à 10, foi por este motivo que utilizamos a variação no parametros destes valores. Também foi utilizado o *early stop* que é outra forma usada para evitar o *overfitting*. É muito utilizado para verificar se o modelo de aprendizado continua aprendendo a cada epoch, caso a acurácia de treinamento caía o treinamento é interropido para que não haja uma piora nos resultados, pois caso o algoritmo continuasse poderia haver o *overfitting*.
Para os *solvers* sgd e adam foram usados a variação das epoch de 100 à 500, com incremento de 50. Já para o *solver* lbfgs foi utilizado a variação de 1000 à 5000, com incremento de 1000. Para o número de camadas foram testados dois valores 1 e 2, já para as camadas escondidas variando de 1 até 100.


Diante dessa configuração, 100 combinações foram avaliadas, sendo a melhor combinação a de número 84 composta com: Número de camadas = 1, camadas escondidas_1 = 79, função de ativação: relu, solver: adam, L2 = 1,78 e número de epochs = 250, neste caso não foi necessário utilizar o Early stopping. Essa combinação obteve uma acurácia de 0,84 para   treinamento e 0,87 para teste.


Observa-se que a acurácia de treinamento e de teste tem um diferença de apenas 3 pontos percentuais, apesar disso, conclui-se que o modelo treinado é adequado e não sofreu *underfitting* e nem *overfitting* sendo útil para tarefa de aprovação de crédito conforme experimentação realizada. Observa-se também que se o algoritmo for treinado com estes parâmetros é possível que ocorra uma melhora da acurácia.

O algoritmo redes neurais apresentou ser um algoritmo poderoso em relação aos outros modelos analisados, pois obteve a acurácia de 0,87 na fase de teste; enquanto os algoritmos SVR, K-NN e LVQ (com os valores ótimos para o conjunto de teste) - experimentados em missões anteriores -  obtiveram acurácia de 0,83, 0,78 e 0,62 respectivamente.



## <center> Referências </center>

* https://stats.stackexchange.com/questions/181/how-to-choose-the-number-of-hidden-layers-and-nodes-in-a-feedforward-neural-netw


* https://medium.datadriveninvestor.com/when-not-to-use-neural-networks-89fb50622429

* https://datascience.eu/pt/inteligencia-artificial/como-funciona-o-algoritmo-de-retropropagacao/

* https://www.kdnuggets.com/2019/12/5-techniques-prevent-overfitting-neural-networks.html

* https://www.deeplearningbook.com.br/overfitting-e-regularizacao-parte-2/

* https://medium.com/turing-talks/turing-talks-20-regress%C3%A3o-de-ridge-e-lasso-a0fc467b5629

* https://medium.com/data-hackers/o-que-%C3%A9-regulariza%C3%A7%C3%A3o-caa1967b8b13
